File: backend/src/main/java/com/lifeos/backend/task/domain/policy/TaskValidationPolicy.java
package com.lifeos.backend.task.domain.policy;

import com.lifeos.backend.task.domain.Task;
import com.lifeos.backend.task.domain.enums.TaskMode;
import com.lifeos.backend.task.domain.enums.TaskRecurrenceType;
import com.lifeos.backend.task.domain.valueobject.TaskRecurrenceRule;
import org.springframework.stereotype.Component;

@Component
public class TaskValidationPolicy {

    public void validate(Task task) {
        validateProgressRules(task.getTaskMode(), task.getProgressPercent());
        validateRecurrence(task.getRecurrenceRule());
        validatePlanningShape(task);
    }

    public void validateProgressRules(TaskMode taskMode, Integer progressPercent) {
        if (progressPercent == null) {
            return;
        }

        if (taskMode != TaskMode.PROGRESS) {
            throw new IllegalArgumentException("Progress percent is only allowed for PROGRESS tasks");
        }

        if (progressPercent < 0 || progressPercent > 100) {
            throw new IllegalArgumentException("Progress percent must be between 0 and 100");
        }
    }
    private void validatePlanningShape(Task task) {
        if (task == null) {
            return;
        }

        boolean hasDueDate = task.getDueDate() != null || task.getDueDateTime() != null;
        boolean isRecurring = task.isRecurring();

        if (isRecurring && hasDueDate) {
            throw new IllegalArgumentException(
                    "Recurring task must not have dueDate or dueDateTime"
            );
        }

        if (isRecurring) {
            if (task.getRecurrenceRule() == null
                    || task.getRecurrenceRule().getStartDate() == null) {
                throw new IllegalArgumentException(
                        "Recurring task must have recurrenceStartDate"
                );
            }
        }
    }
    public void validateRecurrence(TaskRecurrenceRule recurrenceRule) {
        if (recurrenceRule == null) {
            return;
        }

        TaskRecurrenceType type = recurrenceRule.getType();

        if (type == null) {
            return;
        }

        if (type != TaskRecurrenceType.NONE && recurrenceRule.getStartDate() == null) {
            throw new IllegalArgumentException("recurrenceStartDate is required for recurring task");
        }

        if (recurrenceRule.getEndDate() != null
                && recurrenceRule.getStartDate() != null
                && recurrenceRule.getEndDate().isBefore(recurrenceRule.getStartDate())) {
            throw new IllegalArgumentException("recurrenceEndDate must be on or after recurrenceStartDate");
        }

        if (type == TaskRecurrenceType.CUSTOM_WEEKLY
                && (recurrenceRule.getDaysOfWeek() == null || recurrenceRule.getDaysOfWeek().isBlank())) {
            throw new IllegalArgumentException("recurrenceDaysOfWeek is required for CUSTOM_WEEKLY");
        }
    }
}

File: mobile/lib/features/task/domain/policy/task_validation_policy.dart
import '../command/create_task_command.dart';
import '../command/update_task_command.dart';
import '../enum/task_mode.dart';
import '../enum/task_recurrence_type.dart';

class TaskValidationPolicy {
  const TaskValidationPolicy();

  String? validateCreate(CreateTaskCommand command) {
    final titleError = validateTitle(command.title);
    if (titleError != null) return titleError;

    return _validateRules(
      taskMode: command.taskMode,
      progressPercent: command.progressPercent,
      recurrenceType: command.recurrenceType,
      recurrenceStartDate: command.recurrenceStartDate,
      recurrenceEndDate: command.recurrenceEndDate,
      recurrenceDaysOfWeek: command.recurrenceDaysOfWeek,
    );
  }

  String? validateUpdate(UpdateTaskCommand command) {
    if (command.title != null) {
      final titleError = validateTitle(command.title!);
      if (titleError != null) return titleError;
    }

    return _validateRules(
      taskMode: command.taskMode,
      progressPercent: command.progressPercent,
      recurrenceType: command.recurrenceType,
      recurrenceStartDate: command.recurrenceStartDate,
      recurrenceEndDate: command.recurrenceEndDate,
      recurrenceDaysOfWeek: command.recurrenceDaysOfWeek,
    );
  }

  String? validateTitle(String title) {
    if (title.trim().isEmpty) {
      return 'Task title is required';
    }

    return null;
  }

  String? _validateRules({
    required TaskMode? taskMode,
    required int? progressPercent,
    required TaskRecurrenceType? recurrenceType,
    required DateTime? recurrenceStartDate,
    required DateTime? recurrenceEndDate,
    required List<String>? recurrenceDaysOfWeek,
  }) {
    if (progressPercent != null) {
      if (taskMode != TaskMode.progress) {
        return 'Progress percent is only allowed for progress tasks';
      }

      if (progressPercent < 0 || progressPercent > 100) {
        return 'Progress percent must be between 0 and 100';
      }
    }

    if (taskMode == TaskMode.progress &&
        (recurrenceType ?? TaskRecurrenceType.none).isRecurring) {
      return 'Progress tasks represent a finite milestone to 100% and cannot be recurring. Use a Daily task instead.';
    }

    final type = recurrenceType ?? TaskRecurrenceType.none;

    if (type.isRecurring && recurrenceStartDate == null) {
      return 'Recurrence start date is required';
    }

    if (recurrenceStartDate != null &&
        recurrenceEndDate != null &&
        recurrenceEndDate.isBefore(recurrenceStartDate)) {
      return 'Recurrence end date must be on or after start date';
    }

    if (type == TaskRecurrenceType.customWeekly &&
        (recurrenceDaysOfWeek == null || recurrenceDaysOfWeek.isEmpty)) {
      return 'Choose at least one weekday';
    }

    return null;
  }
}
